File: Algorithms/0221_Maximal_Square/Python/Maximal_Square_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def maximalSquare(self, matrix):
        m = len(matrix)
        n = len(matrix[0])
        if m == 1 and n == 1: return 1 if matrix[0][0] == '1' else 0
        dp = [[0 for _ in range(n)] for _ in range(m)]
        res = 0
        for y in range(m):
            for x in range(n):
                if y == 0 or x == 0:
                    dp[y][x] = int(matrix[y][x])
                elif matrix[y][x] == '1':
                    dp[y][x] = min(dp[y][x - 1], dp[y - 1][x - 1], dp[y - 1][x]) + int(matrix[y][x])
                res = max(res, dp[y][x])
        return res * res

File: Algorithms/0221_Maximal_Square/Java/MaximalSquareSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int maximalSquare(char[][] matrix) {
        if (matrix.length == 1 && matrix[0].length == 1) {
            return matrix[0][0] == '1' ? 1 : 0;
        }
        int[][] dp = new int[matrix.length][matrix[0].length];
        for (int[] row : dp) {
            Arrays.fill(row, 0);
        }
        int res = 0;
        for (int y = 0; y < matrix.length; y++) {
            for (int x = 0; x < matrix[0].length; x++) {
                if (y == 0 || x == 0) {
                    dp[y][x] = Character.getNumericValue(matrix[y][x]);
                } else if (matrix[y][x] == '1') {
                    dp[y][x] = Math.min(Math.min(dp[y - 1][x], dp[y - 1][x - 1]), dp[y][x - 1]) + 1;
                }
                res = Math.max(res, dp[y][x]);
            }
        }
        return res * res;
    }
}





